src/main/java/duke/filehandler/Storage.java

```diff
@@ -0,0 +1,122 @@
+        }
+    }
+
+    /***
+     * Saves tasks in memory to file
+     * @param tasks
+     * @throws DukeException
+     */
+    public static void saveTasks(ArrayList<Task> tasks) throws DukeException {
+        try {
+            File taskFile = new File(FILE_PATH);
+            if (!taskFile.exists()) {
+                taskFile.createNewFile();
+            }
+        } catch (IOException e) {
+            throw new DukeException("Error creating task file!");
+        }
+
+        try {
+            FileWriter taskWriter = new FileWriter(FILE_PATH);
+            for(Task task : tasks){
+                taskWriter.write(task.toString());
+                taskWriter.write("\n");
+            }
+            taskWriter.close();
+        } catch (IOException e) {
+            throw new DukeException(" Error saving tasks to file");
+        }
+    }
+
+    /***
+     * reads all tasks from saved file
+     * @return tasks as an ArrayList
+     * @throws DukeException
+     */
+
+    public static ArrayList<Task> readTasks() throws DukeException {
+        ArrayList<Task> tasks = new ArrayList<>();
+
+        try{
+            File taskFile = new File(FILE_PATH);
+            if (!taskFile.exists()){
+                return tasks;
+            }
+            Scanner reader = new Scanner(taskFile);
+            while (reader.hasNextLine()) {
+                String data = reader.nextLine();
+                if (data.length() > 0){
+                    if (data.charAt(1)=='T'){
+                        Todo todo = new Todo (data.substring(7));
+                        if (data.charAt(4) == 'X'){
+                            todo.setAction("mark");
+                        }
+                        tasks.add(todo);
+                    } else if (data.charAt(1)=='D'){
+                        String dates = data.split(":",2)[1];
+                        String desc = data.split("\\(")[0];
+                        Date date = fileDateParser(dates.substring(1,dates.length()-1));
+                        Deadline deadline = new Deadline(desc.substring(7), date);
+                        if (data.charAt(4) == 'X'){
+                            deadline.setAction("mark");
+                        }
+                        tasks.add(deadline);
+                    } else if (data.charAt(1) == 'E'){
+                        String[] dates = data.split(":");
+                        String desc = data.split("\\(")[0];
+                        Date from = fileDateParser(dates[1].substring(1,dates[1].length()-2));
+                        Date to = fileDateParser(dates[2].substring(1,dates[2].length()-1));
+                        Event event = new Event(desc.substring(7), from,to);
+                        if (data.charAt(4) == 'X'){
+                            event.setAction("mark");
+                        }
+                        tasks.add(event);
+                    }
+                }
+            }
+        } catch( IOException e){
```

Comment:
Please add or remove appropriate spacing here, primarily after commas, and around Java keywords and operators, as per our coding standards.

For example, in this segment:
```suggestion
                if (data.length() > 0) {
                    if (data.charAt(1) == 'T') {
                        Todo todo = new Todo(data.substring(7));
                        if (data.charAt(4) == 'X') {
                            todo.setAction("mark");
                        }
                        tasks.add(todo);
                    } else if (data.charAt(1) == 'D') {
                        String dates = data.split(":", 2)[1];
                        String desc = data.split("\\(")[0];
                        Date date = fileDateParser(dates.substring(1, dates.length() - 1));
                        Deadline deadline = new Deadline(desc.substring(7), date);
                        if (data.charAt(4) == 'X') {
                            deadline.setAction("mark");
                        }
                        tasks.add(deadline);
                    } else if (data.charAt(1) == 'E') {
                        String[] dates = data.split(":");
                        String desc = data.split("\\(")[0];
                        Date from = fileDateParser(dates[1].substring(1, dates[1].length() - 2));
                        Date to = fileDateParser(dates[2].substring(1, dates[2].length() - 1));
                        Event event = new Event(desc.substring(7), from, to);
                        if (data.charAt(4) == 'X') {
                            event.setAction("mark");
                        }
                        tasks.add(event);
                    }
                }
            }
        } catch (IOException e) {
```

I also noticed that these formatting issues show up in other areas as well, so do update them appropriately. To save time, you might want to use `codestyle` to automatically detect them, and configure your IDE to automatically format them as well.